blockchain/blockdao/blockdao.go

```diff
@@ -680,20 +681,24 @@ func (dao *blockDAO) deleteTipBlock() error {
 	if err := whichDB.Commit(batchForBlock); err != nil {
 		return err
 	}
+	// Delete hash -> height mapping
+	hashKey := append(hashPrefix, hash[:]...)
+	batch.Delete(blockHashHeightMappingNS, hashKey, "failed to delete hash -> height mapping")
+
+	// Delete height -> hash mapping
+	heightKey := append(heightPrefix, byteutil.Uint64ToBytes(height)...)
+	batch.Delete(blockHashHeightMappingNS, heightKey, "failed to delete height -> hash mapping")
 
 	// Update tip height
-	if err := dao.kvstore.Put(blockNS, topHeightKey, byteutil.Uint64ToBytes(height-1)); err != nil {
-		return errors.Wrap(err, "failed to update top height")
-	}
+	batch.Put(blockNS, topHeightKey, byteutil.Uint64ToBytes(height-1), "failed to put top height")
+
 	// Update tip hash
 	hash, err = dao.getBlockHash(height - 1)
 	if err != nil {
 		return errors.Wrap(err, "failed to get tip block hash")
 	}
-	if err := dao.kvstore.Put(blockNS, topHashKey, hash[:]); err != nil {
-		return errors.Wrap(err, "failed to get tip block hash")
-	}
-	return nil
+	batch.Put(blockNS, topHashKey, hash[:], "failed to put top hash")
+	return dao.kvstore.Commit(batch)
```

Comment:
"whichDB" is supposed to be the "source of truth", which contains the blocks. Thus, we need to delete on kvstore, and then "whichDB", in case of failure.

Reply:
fixed in latest commit